utils/aplay/aplay.c

```diff
@@ -116,6 +127,132 @@ static snd_pcm_format_t bluealsa_get_snd_pcm_format(const struct ba_pcm *pcm) {
+	DBusMessage *msg, *rep = NULL;
+	DBusError err = DBUS_ERROR_INIT;
+
+	msg = dbus_message_new_method_call(dbus_ctx.ba_service,
+					pcm->pcm_path, DBUS_INTERFACE_PROPERTIES, "Set");
+
+	static const char *interface = BLUEALSA_INTERFACE_PCM;
+	static const char *property = "Volume";
+
+	DBusMessageIter iter;
+	DBusMessageIter iter_val;
+
+	dbus_message_iter_init_append(msg, &iter);
+	dbus_message_iter_append_basic(&iter, DBUS_TYPE_STRING, &interface);
+	dbus_message_iter_append_basic(&iter, DBUS_TYPE_STRING, &property);
+	dbus_message_iter_open_container(&iter, DBUS_TYPE_VARIANT, DBUS_TYPE_UINT16_AS_STRING, &iter_val);
+	dbus_message_iter_append_basic(&iter_val, DBUS_TYPE_UINT16, &pcm->volume.raw);
+	dbus_message_iter_close_container(&iter, &iter_val);
+
+	if ((rep = dbus_connection_send_with_reply_and_block(dbus_ctx.conn, msg,
+					DBUS_TIMEOUT_USE_DEFAULT, &err)) == NULL) {
+		warn("Couldn't set volume: %s", err.message);
+	}
+
+	if (msg != NULL)
+		dbus_message_unref(msg);
+	if (rep != NULL)
+		dbus_message_unref(rep);
+}
+
+static void set_transport_volume(struct pcm_worker *w) {
+	struct ba_pcm *ba_pcm = &w->ba_pcm;
+	uint16_t old = ba_pcm->volume.ch1_volume;
+
+	if (ba_pcm->flags & BA_PCM_FLAG_PROFILE_A2DP) {
+		double n_vol;
+		int volume;
+
+		n_vol = get_normalized_playback_volume(w->melem, SND_MIXER_SCHN_FRONT_LEFT);
+		volume = lrint(127.*n_vol);
+
+		debug("ALSA volume (normalized)=%f, volume=%d", n_vol, volume);
+
+		ba_pcm->volume.ch1_volume = volume;
+		if (ba_pcm->channels == 2)
+			ba_pcm->volume.ch2_volume = volume;
+
+		if (ba_pcm->volume.ch1_volume != old) {
+			debug("Setting transport volume to %d (old=%d)", ba_pcm->volume.ch1_volume, old);
+			set_transport_volume_with_dbus(ba_pcm);
+		}
+	}
+}
+
+static int melem_event(snd_mixer_elem_t *elem, unsigned int mask) {
+	size_t i;
+
+	pthread_rwlock_wrlock(&workers_lock);
+
+	for (i = 0; i < workers_count; i++)
+		workers[i].mixer_event = true;
+
+	pthread_rwlock_unlock(&workers_lock);
+
+	return 0;
+};
+
+static snd_mixer_elem_t* init_mixer_control(snd_mixer_t **handle, const char *device, const char *name)
+{
+	snd_mixer_elem_t *elem = NULL;
+	int err;
+	if ((err = snd_mixer_open(handle, 0)) < 0) {
+		error("Mixer open error: %s", snd_strerror(err));
+		goto fail;
+	}
+	if ((err = snd_mixer_attach(*handle, device)) < 0) {
+		error("Mixer attach %s error: %s", device, snd_strerror(err));
+		snd_mixer_close(*handle);
+		goto fail;
+	}
+	if ((err = snd_mixer_selem_register(*handle, NULL, NULL)) < 0) {
+		error("Mixer register error: %s", snd_strerror(err));
+		snd_mixer_close(*handle);
+		goto fail;
+	}
+	if ((err = snd_mixer_load(*handle)) < 0) {
+		error("Mixer %s load error: %s", device, snd_strerror(err));
+		snd_mixer_close(*handle);
+		goto fail;
+	}
+	for (elem = snd_mixer_first_elem(*handle); elem; elem = snd_mixer_elem_next(elem)) {
+		if (!strcmp(snd_mixer_selem_get_name(elem), name))
+			break;
+	}
+	snd_mixer_elem_set_callback(elem, melem_event);
```

Comment:
elem could be NULL here if no element called _name_ exists. so snd_mixer_elem_set_callback() needs to be guarded.

Reply:
Thanks! I will improve the mixer handling.